cpp.mk

```diff
@@ -0,0 +1,181 @@
+# Common C++ Makefile infrastructure
+#
+# Product specific Makefiles define a list of TARGETS and TEST_TARGETS
+# and then `include` this Makefile.  To customise the build product
+# Makefiles can define the following to tweak the behavior:
+#
+#   target_SOURCES  - List the .cpp files to build (defaults to target.cpp only)
+#   target_CPPFLAGS - Specific flags to pass to the C++ compiler
+#   target_LDFLAGS  - Specific flags to pass to the linker
+
+.DEFAULT_GOAL := all
+.PHONY : all test valgrind vg_check coverage_check coverage_raw clean
+
+# Makefiles can override these if needed
+GTEST_DIR ?= ../modules/gmock/gtest
+GMOCK_DIR ?= ../modules/gmock
+CPP_COMMON_DIR ?= ../modules/cpp-common
+BUILD_DIR ?= ../build
+
+# Common rules to build any target
+#
+# @param $1 - Target name (final executable)
+# @param $2 - Target
+define common_target
+
+# Calculate the object names for the $1 build
+$1_OBJS := $$(patsubst %.cpp,$${BUILD_DIR}/$1/%.o,$${$1_SOURCES})
+$1_DEPS := $$(patsubst %.cpp,$${BUILD_DIR}/$1/%.d,$${$1_SOURCES})
+
+# Create alias for the object directory (for adding custom pre-requisites)
+$1_OBJECT_DIR := $${BUILD_DIR}/$1
+
+# Depend on the depends file so we'll get rebuilt if it's missing
+# (which will have the side effect of re-producing the depends file)
+$${$1_OBJS} : $${$1_OBJECT_DIR}/%.o : %.cpp $${BUILD_DIR}/$1/%.d ${MAKEFILE_LIST}
+	@mkdir -p $${$1_OBJECT_DIR}
+	g++ -MMD -MP $${$2_CPPFLAGS} $${$1_CPPFLAGS} -c $$< -o $$@
+
+# Blank rule for depends files to force the object to be re-built
+# if the depends file is missing (as we don't know if it's out of
+# date or not)
+$${$1_DEPS} : $${$1_OBJECT_DIR}/%.d : ;
+
+# Final linker step for $1
+$${BUILD_DIR}/bin/$1 : $${$1_OBJS}
+	@mkdir -p $${BUILD_DIR}/bin/
+	g++ -o $$@ $$^ $${$2_LDFLAGS} $${$1_LDFLAGS}
+
+# Shortcut alias to make $1
+.PHONY : $1
+$1 : $${BUILD_DIR}/bin/$1
+
+# Include depends files from $1
+DEPENDS += $${$1_DEPS}}
+
+# Clean up for $1
+CLEANS += $${$1_OBJS} $${$1_DEPS} $${BUILD_DIR}/bin/$1
+
+endef
+
+# Extra rules for TEST_TARGETS
+#
+# @param $1 - Target name (final executable)
+define test_target
+
+# Call into standard build infrastructure for $1
+$(call common_target,$1,test)
+
+$${BUILD_DIR}/bin/$1 : $${BUILD_DIR}/obj/gmock-all.o $${BUILD_DIR}/obj/gtest-all.o
+
+.PHONY : run_$1
+run_$1 : $${BUILD_DIR}/bin/$1
+	LD_LIBRARY_PATH=../usr/lib/ $$< ${EXTRA_TEST_ARGS}
+
+# This sentinel file proves the tests have *all* been run on this build (mostly for coverage)
+$${BUILD_DIR}/$1/.$1_already_run : $${BUILD_DIR}/bin/$1
+	LD_LIBRARY_PATH=../usr/lib/ $$<
+	@touch $$@
+
+.PHONY : debug_$1
+debug_$1 : $${BUILD_DIR}/bin/$1
+	LD_LIBRARY_PATH=../usr/lib/ gdb --args $$< $${EXTRA_TEST_ARGS}
+
+$${BUILD_DIR}/$1/valgrind_output.xml : $${BUILD_DIR}/bin/$1
+	LD_LIBRARY_PATH=../usr/lib/ valgrind --xml=yes --xml-file=$$@ $$<
```

Comment:
Good spot, I've moved most of that to common and Ralf's Makefile now (optionally) provides a  custom suppression file.